tiledb/sm/query/query.cc

```diff
@@ -281,404 +281,312 @@ Status Query::capnp(rest::capnp::Query::Builder* queryBuilder) {
+        RETURN_NOT_OK(rest::capnp::utils::set_capnp_array_ptr(
+            subarray_builder, domain_type, subarray, subarray_length));
+      }
+      break;
   }
 
-  std::vector<std::string> attributes = this->attributes();
-
-  ::capnp::List<::tiledb::rest::capnp::AttributeBufferHeader>::Builder
-      attributeBufferHeaderBuilder =
-          queryBuilder->initBuffers(attributes.size());
-
-  uint64_t totalNumOfBytesInBuffers = 0;
-  uint64_t totalNumOfBytesInOffsets = 0;
-  uint64_t totalNumOfBytesInVarBuffers = 0;
-  // for (auto attribute_name : attributes) {
-  for (uint64_t i = 0; i < attributes.size(); i++) {
-    std::string attribute_name = attributes[i];
-    ::tiledb::rest::capnp::AttributeBufferHeader::Builder
-        attributeBufferHeader = attributeBufferHeaderBuilder[i];
-    const tiledb::sm::Attribute* attribute =
-        this->array_schema()->attribute(attribute_name);
-    if (attribute != nullptr) {
-      attributeBufferHeader.setAttributeName(attributes[i]);
-      attributeBufferHeader.setType(datatype_str(attribute->type()));
-
-      auto attr_type = attribute->type();
-      auto buff = type_ == QueryType::READ ? reader_.buffer(attribute_name) :
-                                             writer_.buffer(attribute_name);
-
-      if (attribute->var_size() &&
-          (buff.buffer_var_ != nullptr && buff.buffer_var_size_ != nullptr)) {
-        // Variable-sized attribute.
-        if (buff.buffer_ == nullptr || buff.buffer_size_ == nullptr)
-          return Status::RestError(
-              "Cannot serialize query; variable-length attribute has no offset "
-              "buffer set.");
-        totalNumOfBytesInVarBuffers += *buff.buffer_var_size_;
-        attributeBufferHeader.setBufferSizeInBytes(*buff.buffer_var_size_);
-        attributeBufferHeader.setDatatypeSizeInBytes(datatype_size(attr_type));
-        totalNumOfBytesInOffsets += *buff.buffer_size_;
-        attributeBufferHeader.setOffsetBufferSizeInBytes(*buff.buffer_size_);
-        RETURN_NOT_OK(set_buffer(
-            attributes[i],
-            static_cast<uint64_t*>(buff.buffer_),
-            buff.buffer_size_,
-            buff.buffer_var_,
-            buff.buffer_var_size_))
-      } else if (buff.buffer_ != nullptr && buff.buffer_size_ != nullptr) {
-        totalNumOfBytesInBuffers += *buff.buffer_size_;
-        attributeBufferHeader.setBufferSizeInBytes(*buff.buffer_size_);
-        attributeBufferHeader.setDatatypeSizeInBytes(datatype_size(attr_type));
-        attributeBufferHeader.setOffsetBufferSizeInBytes(0);
-        RETURN_NOT_OK(
-            set_buffer(attributes[i], buff.buffer_, buff.buffer_size_))
-      }
+  // Serialize attribute buffer metadata
+  const auto attr_names = attributes();
+  auto attr_buffers_builder =
+      queryBuilder->initAttributeBufferHeaders(attr_names.size());
+  uint64_t total_fixed_len_bytes = 0;
+  uint64_t total_var_len_bytes = 0;
+  for (uint64_t i = 0; i < attr_names.size(); i++) {
```

Comment:
Just to validate, attributes() returns the vector of names based on the user set buffers (query.set_buffer calls), so `__coords` will also be included if set right?

Reply:
Yes.